src/tools/w3dviewer/toolbar.cpp

```diff
@@ -0,0 +1,265 @@
+        }
+
+        if (m_items[i].m_activeBitmap != nullptr) {
+            DeleteObject(m_items[i].m_activeBitmap);
+            m_items[i].m_activeBitmap = nullptr;
+        }
+    }
+}
+
+BOOL CFancyToolbar::PreCreateWindow(CREATESTRUCT &cs)
+{
+    return CControlBar::PreCreateWindow(cs);
+}
+
+CSize CFancyToolbar::CalcFixedLayout(BOOL bStretch, BOOL bHorz)
+{
+    return CSize(42 * m_itemCount + 16, 52);
+}
+
+CSize CFancyToolbar::CalcDynamicLayout(int nLength, DWORD nMode)
+{
+    return CSize(42 * m_itemCount + 16, 52);
+}
+
+BOOL CFancyToolbar::Create(LPCTSTR lpszWindowName, CWnd *pParentWnd, UINT nID)
+{
+    RECT rect;
+    memset(&rect, 0, sizeof(rect));
+
+    BOOL ret = CWnd::Create("FANCYTOOLBAR", lpszWindowName, WS_CHILD | WS_VISIBLE, rect, pParentWnd, nID);
+
+    if (ret) {
+        SetBarStyle(m_dwStyle | CBRS_TOOLTIPS | CBRS_FLYBY | CBRS_SIZE_DYNAMIC);
+        EnableDocking(CBRS_ALIGN_ANY);
+    }
+
+    return ret;
+}
+
+void CFancyToolbar::AddItem(USHORT inactiveBitmapID, USHORT activeBitmapID, UINT id, BOOL activateOnPress)
+{
+    int count = m_itemCount;
+    m_itemCount = count + 1;
+    m_items[count].m_inactiveBitmap = LoadBitmap(AfxGetResourceHandle(), MAKEINTRESOURCE(inactiveBitmapID));
+    m_items[count].m_activeBitmap = LoadBitmap(AfxGetResourceHandle(), MAKEINTRESOURCE(activeBitmapID));
+    m_items[count].m_id = id;
+    m_items[count].m_activateOnPress = activateOnPress;
+    m_items[count].m_isActive = FALSE;
+    m_items[count].m_isVisible = TRUE;
+}
+
+void CFancyToolbar::SetStatus(UINT id, BOOL isActive, BOOL invalidate)
+{
+    BOOL found = FALSE;
+
+    for (int i = 0; i < m_itemCount; i++) {
+        if (found) {
+            break;
+        }
+
+        if (m_items[i].m_id == id) {
+            m_items[i].m_isActive = isActive;
+
+            if (invalidate) {
+                Invalidate();
+                UpdateWindow();
+            }
+        }
+    }
+}
+
+void CFancyToolbar::OnPaint()
+{
+    DoPaint();
+}
+
+void CFancyToolbar::OnLButtonDown(UINT nFlags, CPoint pt)
+{
+    int id = -1;
+    int pos = 8;
+
+    for (int i = 0; i < m_itemCount; i++) {
+        if (id != -1) {
+            break;
+        }
+
+        if (m_items[i].m_isVisible && pt.x >= pos && pt.x <= pos + 42 && pt.y >= 8 && pt.y <= 44) {
+            id = i;
+        }
+
+        pos += 42;
+    }
+
+    if (id >= 0) {
+        m_items[id].m_isActive = !m_items[id].m_isActive;
+
+        HBITMAP bitmap = m_items[id].m_inactiveBitmap;
+
+        if (m_items[id].m_isActive) {
+            bitmap = m_items[id].m_activeBitmap;
+        }
+
+        HDC dc = ::GetDC(m_hWnd);
+
+        if (dc != nullptr) {
+            RECT r;
+            GetClientRect(&r);
+            int cx = ((r.bottom - r.top) >> 1) - 18;
+            HDC dc2 = CreateCompatibleDC(dc);
+
+            if (dc2 != nullptr) {
+                HGDIOBJ obj = SelectObject(dc2, bitmap);
+                BitBlt(dc, 42 * id + 8, cx, 42, 36, dc2, 0, 0, SRCCOPY);
```

Comment:
Can you please add named constants for these magic numbers in this function and class?

Reply:
^